datafusion/physical-plan/src/aggregates/group_values/group_column.rs

```diff
@@ -376,6 +385,399 @@ where
+        //     from last non-inlined `view`
+        //
+        //   - Take buffers, the key is that we need to know if we need to take
+        //     the whole last related buffer. The logic is a bit complex, you can
+        //     detail in `take_buffers_with_whole_last`, `take_buffers_with_partial_last`
+        //     and other related steps in following
+        //
+        //   - Shift the `buffer index` of remaining non-inlined `views`
+        //
+        let first_n_views = self.views.drain(0..n).collect::<Vec<_>>();
+
+        let last_non_inlined_view = first_n_views
+            .iter()
+            .rev()
+            .find(|view| ((**view) as u32) > 12);
+
+        if let Some(view) = last_non_inlined_view {
+            let view = ByteView::from(*view);
+            let last_related_buffer_index = view.buffer_index as usize;
+
+            // Check should we take the whole `last_related_buffer_index` buffer
+            let take_whole_last_buffer = self.should_take_whole_buffer(
+                last_related_buffer_index,
+                (view.offset + view.length) as usize,
+            );
+
+            // Take related buffers
+            let buffers = if take_whole_last_buffer {
+                self.take_buffers_with_whole_last(last_related_buffer_index)
+            } else {
+                self.take_buffers_with_partial_last(
+                    last_related_buffer_index,
+                    (view.offset + view.length) as usize,
+                )
+            };
+
+            // Shift `buffer index`s finally
+            let shifts = if take_whole_last_buffer {
+                last_related_buffer_index + 1
+            } else {
+                last_related_buffer_index
+            };
+
+            self.views.iter_mut().for_each(|view| {
+                if (*view as u32) > 12 {
+                    let mut byte_view = ByteView::from(*view);
+                    byte_view.buffer_index -= shifts as u32;
+                    *view = byte_view.as_u128();
+                }
+            });
+
+            // Build array and return
+            let views = ScalarBuffer::from(first_n_views);
+            Arc::new(GenericByteViewArray::<B>::new(views, buffers, null_buffer))
+        } else {
+            let views = ScalarBuffer::from(first_n_views);
+            Arc::new(GenericByteViewArray::<B>::new(
+                views,
+                Vec::new(),
+                null_buffer,
+            ))
+        }
+    }
+
+    fn take_buffers_with_whole_last(
+        &mut self,
+        last_related_buffer_index: usize,
+    ) -> Vec<Buffer> {
+        if last_related_buffer_index == self.completed.len() {
+            self.flush_in_progress();
+        }
+        self.completed
+            .drain(0..last_related_buffer_index + 1)
+            .collect()
+    }
+
+    fn take_buffers_with_partial_last(
+        &mut self,
+        last_related_buffer_index: usize,
+        take_len: usize,
```

Comment:
maybe we could call this `last_take_len` or something to note it is the number of bytes  being taken from the last buffer